src/isl.rs

```diff
@@ -0,0 +1,211 @@
+use crate::result::{
+    invalid_schema_error_raw, unresolvable_schema_error, IonSchemaError, IonSchemaResult,
+};
+use crate::system::{SharedTypeStore, TypeId};
+use crate::types::TypeDefinition;
+use ion_rs::value::owned::{OwnedElement, OwnedStruct};
+use ion_rs::value::{Element, Sequence, Struct, SymbolToken};
+use ion_rs::IonType;
+use std::convert::{TryFrom, TryInto};
+
+/// Represents a public facing API for schema constraints [IslConstraint] which stores IslTypeRef
+#[derive(Debug, Clone)]
+pub enum IslConstraint {
+    AllOf(Vec<IslTypeRef>),
+    Type(IslTypeRef),
+}
+
+/// Represents a public facing schema type [IslType] which can be converted to a solid TypeRef using TypeStore
+#[derive(Debug, Clone)]
+pub struct IslType {
+    name: Option<String>,
+    constraints: Vec<IslConstraint>,
+}
+
+impl IslType {
+    pub fn new(name: Option<String>, constraints: Vec<IslConstraint>) -> Self {
+        Self { name, constraints }
+    }
+
+    pub fn name(&self) -> &Option<String> {
+        &self.name
+    }
+
+    pub fn constraints(&self) -> &[IslConstraint] {
+        &self.constraints
+    }
+}
+
+/// Parse constraints inside an [OwnedStruct] to an [IslType]
+impl TryFrom<&OwnedStruct> for IslType {
+    type Error = IonSchemaError;
+
+    fn try_from(ion_struct: &OwnedStruct) -> Result<Self, Self::Error> {
+        let mut constraints = vec![];
+
+        // parses the name of the type specified by schema
+        let type_name: Option<String>=  match ion_struct.get("name") {
+            Some(name_element) => match name_element.as_str() {
+                Some(name) => Some(name.to_owned()),
+                None => { return Err(invalid_schema_error_raw("A type name is not string/symbol, if the value is any null, or the text of the symbol is not defined.")) }
+            },
+            None => None // If the type is UNNAMED_TYPE_DEFINITION/ AnonymousType then set name as None
+        };
+
+        // parses all the constraints inside a Type
+        for (field_name, value) in ion_struct.iter() {
+            let constraint_name = match field_name.text() {
+                Some("name") => continue, // if the field_name is "name" then it's the type name not a constraint
+                Some(name) => name,
+                None => {
+                    return Err(invalid_schema_error_raw(
+                        "A type name symbol token does not have any text",
+                    ))
+                }
+            };
+            // TODO: add more constraints to match below
+            let constraint = match constraint_name {
+                "all_of" => {
+                    if value.ion_type() != IonType::List {
+                        return Err(invalid_schema_error_raw(format!(
+                            "all_of constraint was a {:?} instead of a list",
+                            value.ion_type()
+                        )));
+                    }
```

Comment:
question (non-blocking) – Is there a good Rust idiom to DRY some of this IonType checking? I know this only happens a few times in this file right now, but as you implement more of these, you'll find it popping up a lot.

Could we come up with something like this maybe?
```rust
assert_constraint_param_type!("all_of", value, IonType::List)
```

Reply:
Did you mean if we could have  reusable implementation/method for these type checks? I can add a TODO for that and then as I go further with implementation I can use that everywhere.

Reply:
Yes. Sorry that was unclear. (DRY == Don't Repeat Yourself; contrast with WET == Write Everything Twice)

A TODO is just fine.